zebra-state/src/service/tests.rs

```diff
@@ -183,3 +187,87 @@ fn state_behaves_when_blocks_are_committed_out_of_order() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn legacy_chain() -> Result<()> {
+    zebra_test::init();
+
+    legacy_chain_for_network(Network::Mainnet)?;
+    legacy_chain_for_network(Network::Testnet)?;
+
+    Ok(())
+}
+
+/// Test for legacy chain in different scenarios.
+fn legacy_chain_for_network(network: Network) -> Result<()> {
+    zebra_test::init();
+
+    const DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES: u32 = 2;
+    const BLOCKS_AFTER_NU5: u32 = 100;
+
+    if let Some(nu5_height) = NetworkUpgrade::Nu5.activation_height(network) {
+        // Test if we can find at least one transaction with `network_upgrade` field in the chain.
+        let strategy1 = zebra_chain::block::LedgerState::height_strategy(
+            Height(nu5_height.0 + BLOCKS_AFTER_NU5),
+            Some(NetworkUpgrade::Nu5),
+            Some(5),
+            true,
+        )
+        .prop_flat_map(|init| Block::partial_chain_strategy(init, BLOCKS_AFTER_NU5 as usize));
+
+        proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
+            .ok()
+            .and_then(|v| v.parse().ok())
+            .unwrap_or(DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES)),
+            |(chain in strategy1)| {
+                let response = crate::service::legacy_chain_check(Height(nu5_height.0 + BLOCKS_AFTER_NU5), chain.into_iter(), network);
+                prop_assert_eq!(response.is_err(), false);
+                prop_assert_eq!(response.unwrap(), ());
+            }
+        );
+
+        // Test that we can't find at least one transaction with `network_upgrade` field in the chain.
+        let strategy2 = zebra_chain::block::LedgerState::height_strategy(
+            Height(nu5_height.0 + BLOCKS_AFTER_NU5),
+            Some(NetworkUpgrade::Nu5),
+            Some(4),
+            true,
+        )
+        .prop_flat_map(|init| Block::partial_chain_strategy(init, BLOCKS_AFTER_NU5 as usize + 1));
+
+        proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
+            .ok()
+            .and_then(|v| v.parse().ok())
+            .unwrap_or(DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES)),
+            |(chain in strategy2)| {
+                let response = crate::service::legacy_chain_check(Height(nu5_height.0 + BLOCKS_AFTER_NU5 + 1), chain.into_iter(), network);
+                prop_assert_eq!(response.is_err(), true);
+                prop_assert_eq!(response.err().unwrap().to_string(), "giving up after checking too many blocks");
```

Comment:
Minor idea: I think this could also be simplified, but it requires adding a `.map_err(|error| error.to_string())` to the `response`:

```suggestion
                let response = crate::service::legacy_chain_check(
                    Height(nu5_height.0 + BLOCKS_AFTER_NU5 + 1),
                    chain.into_iter(),
                    network,
                ).map_err(|error| error.to_string());
                
                prop_assert_eq!(response.is_err(), Err();
                prop_assert_eq!(response, Err("giving up after checking too many blocks".into()));
```

Reply:
Do not apply anymore.